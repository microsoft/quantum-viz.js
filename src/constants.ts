// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

// SVG Namespace
export const svgNS = 'http://www.w3.org/2000/svg';

// Display attributes
/** Left padding of SVG. */
export const leftPadding = 20;
/** x coordinate for first operation on each register. */
export const startX = 80;
/** y coordinate of first register. */
export const startY = 40;
/** Minimum width of each gate. */
export const minGateWidth = 40;
/** Height of each gate. */
export const gateHeight = 40;
/** Padding on each side of gate. */
export const gatePadding = 10;
/** Padding on each side of gate label. */
export const labelPadding = 10;
/** Height between each qubit register. */
export const registerHeight: number = gateHeight + gatePadding * 2;
/** Height between classical registers. */
export const classicalRegHeight: number = gateHeight;
/** Group box inner padding. */
export const groupBoxPadding = gatePadding;
/** Padding between nested groups. */
export const nestedGroupPadding = 2;
/** Additional offset for control button. */
export const controlBtnOffset = 40;
/** Control button radius. */
export const controlBtnRadius = 15;
/** Default font size for gate labels. */
export const labelFontSize = 14;
/** Default font size for gate arguments. */
export const argsFontSize = 12;
/** Starting x coord for each register wire. */
export const regLineStart = 40;

// Panel
/** Panel width */
export const panelWidth = 144;
/** Gap between gates in Add Panel */
export const gateGap = 50;
