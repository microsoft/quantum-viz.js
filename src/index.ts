// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { Sqore } from './sqore';
import { Circuit } from './circuit';
import { StyleConfig } from './styles';

/**
 * Render `circuit` into `container` at the specified layer depth.
 *
 * @param circuit Circuit to be visualized.
 * @param container HTML element for rendering visualization into.
 * @param style Custom visualization style.
 * @param renderDepth Initial layer depth at which to render gates.
 * @param isEditable Optional value enabling/disabling editable feature
 * @param onCircuitChange Optional function to trigger when changing elements in circuit
 */
export const draw = (
    circuit: Circuit,
    container: HTMLElement,
    style: StyleConfig | string = {},
    renderDepth = 0,
    isEditable?: boolean,
    onCircuitChange?: (circuit: Circuit) => void,
): void => {
    const sqore = new Sqore(circuit, style);

    sqore.draw(container, renderDepth, isEditable, onCircuitChange);
};

export { STYLES } from './styles';

// Export types
export type { StyleConfig } from './styles';
export type { Circuit, Qubit, Operation } from './circuit';
