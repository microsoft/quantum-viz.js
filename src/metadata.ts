// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { DataAttributes } from './circuit';

/**
 * Enum for the various gate operations handled.
 */
export enum GateType {
    /** Measurement gate. */
    Measure,
    /** CNOT gate. */
    Cnot,
    /** SWAP gate. */
    Swap,
    /** Single/multi qubit unitary gate. */
    Unitary,
    /** Single/multi controlled unitary gate. */
    ControlledUnitary,
    /** Nested group of classically-controlled gates. */
    ClassicalControlled,
    /** Group of nested gates */
    Group,
    /** Invalid gate. */
    Invalid,
}

/**
 * Metadata used to store information pertaining to a given
 * operation for rendering its corresponding SVG.
 */
export interface Metadata {
    /** Gate type. */
    type: GateType;
    /** Centre x coord for gate position. */
    x: number;
    /** Array of y coords of control registers. */
    controlsY: number[];
    /** Array of y coords of target registers.
     *  For `GateType.Unitary` or `GateType.ControlledUnitary`, this is an array of groups of
     *  y coords, where each group represents a unitary box to be rendered separately.
     */
    targetsY: (number | number[])[];
    /** Gate label. */
    label: string;
    /** Gate arguments as string. */
    displayArgs?: string;
    /** Gate width. */
    width: number;
    /** Children operations as part of group. */
    children?: (Metadata | Metadata[])[];
    /** Custom data attributes to attach to gate element. */
    dataAttributes?: DataAttributes;
}
