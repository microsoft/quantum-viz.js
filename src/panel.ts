import range from 'lodash/range';
import { Operation } from './circuit';
import { _equivOperation } from './draggable';
import { Register } from './register';
import { Sqore } from './sqore';

const extensionPanel = (container: HTMLElement, sqore: Sqore, useRender: () => void): void => {
    const elems = container.querySelectorAll<SVGElement>('[data-id]');
    elems.forEach((elem) =>
        elem.addEventListener('mousedown', () => {
            const dataId = elem.getAttribute('data-id');
            const operation = _equivOperation(dataId, sqore.circuit.operations);
            const newPanelElem = _panel(qubitSize, dispatch, operation || undefined);
            container.replaceChild(newPanelElem, panelElem);
            panelElem = newPanelElem;
        }),
    );
    const dispatch = reducer(container, sqore, useRender);
    const qubitSize = sqore.circuit.qubits.length;
    let panelElem = _panel(qubitSize, dispatch);
    const prevPanelElem = document.querySelector('.panel');
    prevPanelElem ? container.replaceChild(panelElem, prevPanelElem) : container.prepend(panelElem);
};

interface Action {
    type: string;
    payload: Register[];
}

const reducer =
    (container: HTMLElement, sqore: Sqore, useRender: () => void) =>
    (initial: Operation | undefined, action: Action) => {
        if (initial == null) return;

        switch (action.type) {
            case 'TARGET': {
                initial.targets = action.payload;
                break;
            }
            case 'CONTROLS': {
                initial.controls = action.payload;
                break;
            }
        }
        useRender();
    };

const _panel = (qubitSize: number, dispatch: Dispatch, operation?: Operation) => {
    const options = range(qubitSize).map((i) => ({ value: `${i}`, text: `q${i}` }));
    const target = operation?.targets[0].qId;
    const controls = operation?.controls?.map((control) => control.qId);

    const panelElem = _elem('div');
    panelElem.className = 'panel';
    _children(panelElem, [
        _select('Target', 'target-input', options, target || 0, dispatch, operation),
        _checkboxes('Controls', 'controls-input', options, controls || [], dispatch, operation),
        _text('Display', 'display-input', 'display-arg'),
    ]);

    return panelElem;
};

const _elem = (tag: string): HTMLElement => document.createElement(tag);

/**
 * Append all child elements to a parent element
 */
const _children = (parentElem: HTMLElement, childElems: HTMLElement[]) => {
    childElems.map((elem) => parentElem.appendChild(elem));
    return parentElem;
};

interface Option {
    value: string;
    text: string;
}

interface Dispatch {
    (initial: Operation | undefined, action: Action): void;
}

const _select = (
    label: string,
    className: string,
    options: Option[],
    selectedIndex: number,
    dispatch: Dispatch,
    operation?: Operation,
): HTMLElement => {
    const optionElems = options.map(({ value, text }) => _option(value, text));
    const selectElem = _elem('select') as HTMLSelectElement;
    _children(selectElem, optionElems);
    operation == null && selectElem.setAttribute('disabled', 'true');
    selectElem.selectedIndex = selectedIndex;

    const labelElem = _elem('label') as HTMLLabelElement;
    labelElem.textContent = label;

    const divElem = _elem('div') as HTMLDivElement;
    divElem.className = className;
    _children(divElem, [labelElem, selectElem]);

    selectElem.onchange = () => {
        dispatch(operation, { type: 'TARGET', payload: [{ qId: parseInt(selectElem.value) }] });
    };

    return divElem;
};

const _option = (value: string, text: string) => {
    const elem = _elem('option') as HTMLOptionElement;
    elem.value = value;
    elem.textContent = text;
    return elem;
};

const _checkboxes = (
    label: string,
    className: string,
    options: Option[],
    selectedIndexes: number[],
    dispatch: Dispatch,
    operation?: Operation,
) => {
    const checkboxElems = options.map((option, index) => {
        const elem = _checkbox(option.value, option.text);
        const inputElem = elem.querySelector('input') as HTMLInputElement;
        selectedIndexes.includes(index) && inputElem.setAttribute('checked', 'true');
        operation == null && inputElem.setAttribute('disabled', 'true');

        inputElem.onchange = () => {
            const checkedElems = Array.from(divElem.querySelectorAll<HTMLInputElement>('input:checked'));
            const newControls = checkedElems.map((elem) => ({
                qId: parseInt(elem.value),
            }));
            dispatch(operation, { type: 'CONTROLS', payload: newControls });
        };

        return elem;
    });

    const labelElem = _elem('label');
    labelElem.textContent = label;

    const divElem = _elem('div') as HTMLDivElement;
    divElem.className = className;
    _children(divElem, [labelElem, ...checkboxElems]);

    return divElem;
};

const _checkbox = (value: string, text: string) => {
    const inputElem = _elem('input') as HTMLInputElement;
    inputElem.type = 'checkbox';
    inputElem.value = value;

    const labelElem = _elem('label') as HTMLLabelElement;
    labelElem.textContent = text;
    labelElem.prepend(inputElem);
    return labelElem;
};

const _text = (label: string, className: string, value: string) => {
    const labelElem = _elem('label') as HTMLLabelElement;
    labelElem.textContent = label;

    const textElem = _elem('input') as HTMLInputElement;
    textElem.type = 'text';
    textElem.value = value;

    const divElem = _elem('div');
    divElem.className = className;
    _children(divElem, [labelElem, textElem]);

    return divElem;
};

export { extensionPanel };
