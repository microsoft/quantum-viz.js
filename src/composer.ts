import { formatInputs } from './formatters/inputFormatter';
import { formatGates } from './formatters/gateFormatter';
import { formatRegisters } from './formatters/registerFormatter';
import { processOperations } from './process';
import { Circuit } from './circuit';
import { Metadata, GateType } from './metadata';
import { StyleConfig, style, STYLES } from './styles';
import { createUUID } from './utils';

/**
 * Custom JavaScript code to be injected into visualization HTML string.
 * Handles interactive elements, such as classically-controlled operations.
 */
export const addGateClickHandlers = (container: Element | null) : Element | null => {
    container?.querySelectorAll('.classically-controlled-btn').forEach((btn) => {
        // Zoom in on clicked gate
        btn.addEventListener('click', (evt: Event) => {
            const textSvg = btn.querySelector('text');
            const group = btn.parentElement;
            if (textSvg == null || group == null) return;

            const currValue = textSvg.firstChild?.nodeValue;
            const zeroGates = group?.querySelector('.classically-controlled-zero');
            const oneGates = group?.querySelector('.classically-controlled-one');
            switch (currValue) {
                case '?':
                    textSvg.childNodes[0].nodeValue = '1';
                    group.classList.remove('classically-controlled-unknown');
                    group.classList.remove('classically-controlled-zero');
                    group.classList.add('classically-controlled-one');
                    zeroGates?.classList.add('hidden');
                    oneGates?.classList.remove('hidden');
                    break;
                case '1':
                    textSvg.childNodes[0].nodeValue = '0';
                    group.classList.remove('classically-controlled-unknown');;
                    group.classList.add('classically-controlled-zero');
                    group.classList.remove('classically-controlled-one');
                    zeroGates?.classList.remove('hidden');
                    oneGates?.classList.add('hidden');
                    break;
                case '0':
                    textSvg.childNodes[0].nodeValue = '?';
                    group.classList.add('classically-controlled-unknown');
                    group.classList.remove('classically-controlled-zero');
                    group.classList.remove('classically-controlled-one');
                    zeroGates?.classList.remove('hidden');
                    oneGates?.classList.remove('hidden');
                    break;
            }
            evt.stopPropagation();
        });
    });

    return container;
}


/**
 * Contains all metadata required to generate final output.
 */
class ComposedCircuit {
    width: number;
    height: number;
    style: StyleConfig;
    elements: string[];

    /**
     * Initializes `ComposedCircuit` with metadata required for visualization.
     *
     * @param width Width of SVG element.
     * @param height Height of SVG element.
     * @param style Visualization style.
     * @param elements SVG elements the make up circuit visualization.
     */
    constructor(width: number, height: number, style: StyleConfig, elements: string[]) {
        this.width = width;
        this.height = height;
        this.style = style;
        this.elements = elements;
    }

    /**
     * Generates visualization as an SVG string and optionally injects the custom script into the browser.
     *
     * @param injectScript Injects custom script into document manually. This is used when the visualization
     *                     is injected into the document via `innerHtml`.
     *
     * @returns SVG representation of circuit visualization.
     */
    asSvg(): string {
        const uuid: string = createUUID();
        let script = '';

        return `<svg xmlns="http://www.w3.org/2000/svg" version="1.1" id="${uuid}" class="qviz" width="${this.width}" height="${this.height}">
    ${style(this.style)}
    ${this.elements.join('\n')}
</svg>`;
    }

    /**
     * Generates visualization as an HTML string and optionally injects the custom script into the browser.
     *
     * @returns HTML representation of circuit visualization.
     */
    asHtml(): string {
        const svg: string = this.asSvg();
        return `<html>
    ${svg}
</html>`;
    }
}

/**
 * Entrypoint class for rendering circuit visualizations.
 */
class ExecutionPathVisualizer {
    style: StyleConfig;

    /**
     * Initializes ExecutionPathVisualizer object with custom styles.
     *
     * @param style Custom styles for visualization.
     */
    constructor(style: StyleConfig = {}) {
        this.style = style;
    }

    /**
     * Sets custom style for visualization.
     *
     * @param style Custom `StyleConfig` for visualization.
     */
    stylize(style: StyleConfig | string = {}): ExecutionPathVisualizer {
        if (typeof style === 'string' || style instanceof String) {
            const styleName: string = style as string;
            if (!STYLES.hasOwnProperty(styleName)) {
                console.error(`No style ${styleName} found in STYLES.`);
                return this;
            }
            style = STYLES[styleName] || {};
        }
        this.style = style;
        return this;
    }

    /**
     * Generates the components required for visualization.
     *
     * @param circuit Circuit to be visualized.
     *
     * @returns `ComposedCircuit` object containing metadata for visualization.
     */
    compose(circuit: Circuit): ComposedCircuit {
        const flatten = (gates: Metadata[]): Metadata[] => {
            const result: Metadata[] = [];
            function add(acc: Metadata[], gate: Metadata | Metadata[]) {
                if (Array.isArray(gate)) {
                    gate.forEach((g) => add(acc, g));
                } else {
                    acc.push(gate);
                    gate.children?.forEach((g) => add(acc, g));
                }
            }
            add(result, gates);
            return result;
        };

        const { qubits, operations } = circuit;
        const { qubitWires, registers, svgHeight } = formatInputs(qubits);
        const { metadataList, svgWidth } = processOperations(operations, registers);
        const formattedGates: string = formatGates(metadataList);
        const measureGates: Metadata[] = flatten(metadataList).filter(({ type }) => type === GateType.Measure);
        const formattedRegs: string = formatRegisters(registers, measureGates, svgWidth);

        const composition: ComposedCircuit = new ComposedCircuit(svgWidth, svgHeight, this.style, [
            qubitWires,
            formattedRegs,
            formattedGates,
        ]);
        return composition;
    }
}

/** Exported function for creating a new ExecutionPathVisualizer class. */
export const createExecutionPathVisualizer = (): ExecutionPathVisualizer => new ExecutionPathVisualizer();
export { STYLES } from './styles';
export { ConditionalRender } from './circuit';

// Export types
export type { Circuit, StyleConfig, ExecutionPathVisualizer, ComposedCircuit };
export type { Qubit, Operation } from './circuit';
