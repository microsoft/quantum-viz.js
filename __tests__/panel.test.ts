import range from 'lodash/range';
import { PanelContext, PanelOptions } from '../src/panel';
import { Action, Dispatch, exportedForTesting } from '../src/panel';

const {
    panel,
    addPanel,
    editPanel,
    elem,
    children,
    childrenSvg,
    title,
    select,
    option,
    checkboxes,
    checkbox,
    text,
    toMetadata,
    gate,
    defaultGateDictionary,
} = exportedForTesting;

describe('Test elem', () => {
    test('Should return <div>', () => {
        expect(elem('div')).toMatchSnapshot();
    });
    test('Should return <p>', () => {
        expect(elem('p')).toMatchSnapshot();
    });
    test('Should return <div> with className "panel"', () => {
        expect(elem('div', 'panel')).toMatchSnapshot();
    });
});

describe('Test children', () => {
    let parentElem: HTMLDivElement;
    beforeEach(() => {
        parentElem = document.createElement('div');
    });
    test('Add 1 child to parent', () => {
        const childElems = [document.createElement('p')] as HTMLElement[];
        expect(children(parentElem, childElems)).toMatchSnapshot();
    });
    test('Add 2 children to parent', () => {
        const childElems = [document.createElement('p'), document.createElement('div')] as HTMLElement[];
        expect(children(parentElem, childElems)).toMatchSnapshot();
    });
    test('Add 0 with class to parent', () => {
        const childElems = [] as HTMLElement[];
        expect(children(parentElem, childElems)).toMatchSnapshot();
    });
});

describe('Test childrenSvg', () => {
    let parentElem: SVGElement;
    beforeEach(() => {
        parentElem = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
    });
    test('Add 0 child to parent', () => {
        const childElems = [] as SVGElement[];
        expect(childrenSvg(parentElem, childElems)).toMatchSnapshot();
    });
    test('Add 1 child to parent', () => {
        const childElems = [document.createElementNS('http://www.w3.org/2000/svg', 'rect')] as SVGElement[];
        expect(childrenSvg(parentElem, childElems)).toMatchSnapshot();
    });
    test('Add 2 children to parent', () => {
        const childElems = [
            document.createElementNS('http://www.w3.org/2000/svg', 'rect'), //
            document.createElementNS('http://www.w3.org/2000/svg', 'circle'), //
        ] as SVGElement[];
        expect(childrenSvg(parentElem, childElems)).toMatchSnapshot();
    });
});

describe('Test title', () => {
    test('Should return title "PANEL"', () => {
        expect(title('PANEL')).toMatchSnapshot();
    });
    test('Should return title "ADD"', () => {
        expect(title('ADD')).toMatchSnapshot();
    });
    test('Should return title "EDIT"', () => {
        expect(title('EDIT')).toMatchSnapshot();
    });
});

describe('Test text', () => {
    const emptyDispatch: Dispatch = (action: Action) => {
        action;
    };
    test('Should return gate H without display-args', () => {
        const operation = {
            gate: 'H',
            targets: [{ qId: 0 }],
        };
        expect(text('Display', 'display-input', emptyDispatch, operation)).toMatchSnapshot();
    });
    test('Should return gate RX with display-args', () => {
        const operation = {
            gate: 'RX',
            displayArgs: '(0.25)',
            isControlled: true,
            controls: [{ qId: 1 }],
            targets: [{ qId: 0 }],
        };
        expect(text('Display', 'display-input', emptyDispatch, operation)).toMatchSnapshot();
    });
});

describe('Test toMetadata', () => {
    test('Should return metadata of gate H', () => {
        const operation = {
            gate: 'H',
            targets: [{ qId: 0 }],
        };
        expect(toMetadata(operation, 0, 0)).toMatchSnapshot();
    });
    test('Should return metadata of gate RX', () => {
        const operation = {
            gate: 'H',
            targets: [{ qId: 0 }],
        };
        expect(toMetadata(operation, 0, 0)).toMatchSnapshot();
    });
    test('Should return metadata of gate X', () => {
        const operation = {
            gate: 'X',
            targets: [{ qId: 3 }],
        };
        expect(toMetadata(operation, 0, 0)).toMatchSnapshot();
    });
});

describe('Test select', () => {
    const emptyDispatch: Dispatch = (action: Action) => {
        action;
    };
    test('Test with gate X', () => {
        const operation = {
            gate: 'X',
            targets: [{ qId: 0 }],
            controls: [{ qId: 1 }],
        };
        const registerSize = 2;
        const options = range(registerSize).map((i) => ({ value: `${i}`, text: `q${i}` }));
        const target = operation?.targets[0].qId;
        select('Target', 'target-input', options, target || 0, emptyDispatch, operation);
    });
    test('Test with gate H', () => {
        const operation = {
            gate: 'H',
            targets: [{ qId: 0 }],
        };
        const registerSize = 2;
        const options = range(registerSize).map((i) => ({ value: `${i}`, text: `q${i}` }));
        const target = operation?.targets[0].qId;
        select('Target', 'target-input', options, target || 0, emptyDispatch, operation);
    });
});

describe('Test option', () => {
    test('Should return option q0', () => {
        expect(option('0', 'q0')).toMatchSnapshot();
    });
    test('Should return option q1', () => {
        expect(option('1', 'q1')).toMatchSnapshot();
    });
    test('Should return option q2', () => {
        expect(option('2', 'q2')).toMatchSnapshot();
    });
});

describe('Test checkboxes', () => {
    const emptyDispatch: Dispatch = (action: Action) => {
        action;
    };
    test('Test with gate X with 1 control', () => {
        const operation = {
            gate: 'X',
            targets: [{ qId: 0 }],
            controls: [{ qId: 1 }],
        };
        const registerSize = 2;
        const options = range(registerSize).map((i) => ({ value: `${i}`, text: `q${i}` }));
        const controls = operation.controls?.map((control) => control.qId);
        expect(
            checkboxes('Controls', 'controls-input', options, controls || [], emptyDispatch, operation),
        ).toMatchSnapshot();
    });
    test('Test with gate X with 2 controls', () => {
        const operation = {
            gate: 'X',
            targets: [{ qId: 1 }],
            controls: [{ qId: 0 }, { qId: 2 }],
        };
        const registerSize = 3;
        const options = range(registerSize).map((i) => ({ value: `${i}`, text: `q${i}` }));
        const controls = operation.controls?.map((control) => control.qId);
        expect(
            checkboxes('Controls', 'controls-input', options, controls || [], emptyDispatch, operation),
        ).toMatchSnapshot();
    });
});

describe('Test checkbox', () => {
    test('Should return checkbox q0', () => {
        expect(checkbox('0', 'q0')).toMatchSnapshot();
    });
    test('Should return checkbox q1', () => {
        expect(checkbox('1', 'q1')).toMatchSnapshot();
    });
    test('Should return checkbox q2', () => {
        expect(checkbox('2', 'q2')).toMatchSnapshot();
    });
});

describe('Test gate', () => {
    const emptyDispatch: Dispatch = (action: Action) => {
        action;
    };
    test('Should return gate H', () => {
        expect(gate(emptyDispatch, defaultGateDictionary, 'H', 0, 0)).toMatchSnapshot();
    });
    test('Should return gate X', () => {
        expect(gate(emptyDispatch, defaultGateDictionary, 'X', 0, 0)).toMatchSnapshot();
    });
    test('Should return gate RX', () => {
        expect(gate(emptyDispatch, defaultGateDictionary, 'RX', 0, 0)).toMatchSnapshot();
    });
});

describe('Test defaulGateDictionary', () => {
    test('Verify defaultGateDictionary', () => {
        const expected = {
            Entangle: {
                gate: 'Entangle',
                targets: [{ qId: 0 }],
            },
            RX: {
                gate: 'RX',
                targets: [{ qId: 0 }],
            },
            RY: {
                gate: 'RY',
                targets: [{ qId: 0 }],
            },
            RZ: {
                gate: 'RZ',
                targets: [{ qId: 0 }],
            },
            H: {
                gate: 'H',
                targets: [{ qId: 0 }],
            },
            X: {
                gate: 'X',
                targets: [{ qId: 0 }],
            },
            S: {
                gate: 'S',
                targets: [{ qId: 0 }],
            },
            T: {
                gate: 'T',
                targets: [{ qId: 0 }],
            },
            Y: {
                gate: 'Y',
                targets: [{ qId: 0 }],
            },
            Z: {
                gate: 'Z',
                targets: [{ qId: 0 }],
            },
            ZZ: {
                gate: 'ZZ',
                targets: [{ qId: 0 }],
            },
        };
        expect(defaultGateDictionary).toMatchObject(expected);
    });
});

describe('Test editPanel', () => {
    const emptyDispatch: Dispatch = (action: Action) => {
        action;
    };
    test('Should return editPanel editing X gate', () => {
        const context: PanelContext = {
            addMode: false,
            operations: [],
            operation: {
                gate: 'X',
                targets: [{ qId: 0 }],
            },
            registerSize: 2,
            container: undefined,
        };
        expect(editPanel(emptyDispatch, context)).toMatchSnapshot();
    });
});

describe('Test addPanel', () => {
    const emptyDispatch: Dispatch = (action: Action) => {
        action;
    };
    test('Should return default addPanel', () => {
        const context: PanelContext = {
            addMode: true,
            operations: [],
            operation: {
                gate: 'X',
                targets: [{ qId: 0 }],
            },
            registerSize: 2,
            container: undefined,
        };
        expect(addPanel(emptyDispatch, context)).toMatchSnapshot();
    });
    test('Should return default addPanel with displaySize 2', () => {
        const context: PanelContext = {
            addMode: true,
            operations: [],
            operation: {
                gate: 'X',
                targets: [{ qId: 0 }],
            },
            registerSize: 2,
            container: undefined,
        };
        const options: PanelOptions = { displaySize: 2 };
        expect(addPanel(emptyDispatch, context, options)).toMatchSnapshot();
    });
});

describe('Test panel', () => {
    const emptyDispatch: Dispatch = (action: Action) => {
        action;
    };
    test('Should return panel with addPanel visible', () => {
        const context: PanelContext = {
            addMode: true,
            operations: [],
            operation: {
                gate: 'X',
                targets: [{ qId: 0 }],
            },
            registerSize: 2,
            container: undefined,
        };
        expect(panel(emptyDispatch, context)).toMatchSnapshot();
    });
    test('Should return panel with editPanel visible', () => {
        const context: PanelContext = {
            addMode: false,
            operations: [],
            operation: {
                gate: 'X',
                targets: [{ qId: 0 }],
            },
            registerSize: 2,
            container: undefined,
        };
        expect(panel(emptyDispatch, context)).toMatchSnapshot();
    });
});
