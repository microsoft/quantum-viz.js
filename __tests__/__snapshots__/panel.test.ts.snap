// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test checkbox Should return checkbox q0 1`] = `
<label>
  <input
    type="checkbox"
    value="0"
  />
  q0
</label>
`;

exports[`Test checkbox Should return checkbox q1 1`] = `
<label>
  <input
    type="checkbox"
    value="1"
  />
  q1
</label>
`;

exports[`Test checkbox Should return checkbox q2 1`] = `
<label>
  <input
    type="checkbox"
    value="2"
  />
  q2
</label>
`;

exports[`Test checkboxes Test with gate X with 1 control 1`] = `
<div
  class="controls-input"
>
  <label
    class="block"
  >
    Controls
  </label>
  <label>
    <input
      type="checkbox"
      value="0"
    />
    q0
  </label>
  <label>
    <input
      checked="true"
      type="checkbox"
      value="1"
    />
    q1
  </label>
</div>
`;

exports[`Test checkboxes Test with gate X with 2 controls 1`] = `
<div
  class="controls-input"
>
  <label
    class="block"
  >
    Controls
  </label>
  <label>
    <input
      checked="true"
      type="checkbox"
      value="0"
    />
    q0
  </label>
  <label>
    <input
      type="checkbox"
      value="1"
    />
    q1
  </label>
  <label>
    <input
      checked="true"
      type="checkbox"
      value="2"
    />
    q2
  </label>
</div>
`;

exports[`Test children Add 0 with class to parent 1`] = `<div />`;

exports[`Test children Add 1 child to parent 1`] = `
<div>
  <p />
</div>
`;

exports[`Test children Add 2 children to parent 1`] = `
<div>
  <p />
  <div />
</div>
`;

exports[`Test childrenSvg Add 0 child to parent 1`] = `<svg />`;

exports[`Test childrenSvg Add 1 child to parent 1`] = `
<svg>
  <rect />
</svg>
`;

exports[`Test childrenSvg Add 2 children to parent 1`] = `
<svg>
  <rect />
  <circle />
</svg>
`;

exports[`Test editPanel  1`] = `
<div
  class="edit-panel"
>
  <h2
    class="title"
  >
    EDIT
  </h2>
  <div
    class="target-input"
  >
    <label
      class="block"
    >
      Target
    </label>
    <select>
      <option
        value="0"
      >
        q0
      </option>
      <option
        value="1"
      >
        q1
      </option>
    </select>
  </div>
  <div
    class="controls-input"
  >
    <label
      class="block"
    >
      Controls
    </label>
    <label>
      <input
        type="checkbox"
        value="0"
      />
      q0
    </label>
    <label>
      <input
        type="checkbox"
        value="1"
      />
      q1
    </label>
  </div>
  <div
    class="display-input"
  >
    <label
      class="block"
    >
      Display
    </label>
    <input
      autofocus="true"
      type="text"
    />
  </div>
</div>
`;

exports[`Test elem Should return <div> 1`] = `<div />`;

exports[`Test elem Should return <div> with className "panel" 1`] = `
<div
  class="panel"
/>
`;

exports[`Test elem Should return <p> 1`] = `<p />`;

exports[`Test gate Should return gate H 1`] = `
<g
  class="gate"
>
  <g>
    <g>
      <rect
        class="gate-unitary"
        height="40"
        width="40"
        x="1"
        y="1"
      />
      <text
        font-size="14"
        x="21"
        y="21"
      >
        H
      </text>
    </g>
  </g>
</g>
`;

exports[`Test gate Should return gate RX 1`] = `
<g
  class="gate"
>
  <g>
    <g>
      <rect
        class="gate-unitary"
        height="40"
        width="40"
        x="1"
        y="1"
      />
      <text
        font-size="14"
        x="21"
        y="21"
      >
        RX
      </text>
    </g>
  </g>
</g>
`;

exports[`Test gate Should return gate X 1`] = `
<g
  class="gate"
>
  <g
    class="oplus"
  >
    <circle
      cx="21"
      cy="21"
      r="15"
    />
    <line
      x1="21"
      x2="21"
      y1="6"
      y2="36"
    />
    <line
      x1="6"
      x2="36"
      y1="21"
      y2="21"
    />
  </g>
</g>
`;

exports[`Test option Should return option q0 1`] = `
<option
  value="0"
>
  q0
</option>
`;

exports[`Test option Should return option q1 1`] = `
<option
  value="1"
>
  q1
</option>
`;

exports[`Test option Should return option q2 1`] = `
<option
  value="2"
>
  q2
</option>
`;

exports[`Test text Should return gate H without display-args 1`] = `
<div
  class="display-input"
>
  <label
    class="block"
  >
    Display
  </label>
  <input
    autofocus="true"
    type="text"
  />
</div>
`;

exports[`Test text Should return gate RX with display-args 1`] = `
<div
  class="display-input"
>
  <label
    class="block"
  >
    Display
  </label>
  <input
    autofocus="true"
    type="text"
  />
</div>
`;

exports[`Test title Should return title "ADD" 1`] = `
<h2
  class="title"
>
  ADD
</h2>
`;

exports[`Test title Should return title "EDIT" 1`] = `
<h2
  class="title"
>
  EDIT
</h2>
`;

exports[`Test title Should return title "PANEL" 1`] = `
<h2
  class="title"
>
  PANEL
</h2>
`;

exports[`Test toMetadata Should return metadata of gate H 1`] = `
Object {
  "controlsY": Array [],
  "label": "H",
  "targetsY": Array [
    Array [
      21,
    ],
  ],
  "type": 4,
  "width": 40,
  "x": 21,
}
`;

exports[`Test toMetadata Should return metadata of gate RX 1`] = `
Object {
  "controlsY": Array [],
  "label": "H",
  "targetsY": Array [
    Array [
      21,
    ],
  ],
  "type": 4,
  "width": 40,
  "x": 21,
}
`;

exports[`Test toMetadata Should return metadata of gate X 1`] = `
Object {
  "controlsY": Array [],
  "label": "X",
  "targetsY": Array [
    21,
  ],
  "type": 3,
  "width": 40,
  "x": 21,
}
`;
